### ISSUE TITLE
confused by snakemake's grammar and lack of concrete reference material

### ISSUE THREAD

**Issue by grahamgower** [snakemake/snakemake#1168]
I'm finding it hard to understand what is allowed where in a Snakefile. I know both `make` and `Python` reasonably well, and read through most of the snakemake docs. But I'm having a hard time seeing how it all fits together with the grammar section (https://snakemake.readthedocs.io/en/stable/snakefiles/writing_snakefiles.html). My confusion arises at least partly because the grammar is incomplete, and it doesn't really serve as a language reference anyhow. These are the things I'm particularly confused about:

* What's a `parameter_list`? E.g., I would like to know concretely what is allowed in an `input: ` line/section. The grammar indicates this is a `parameter_list`, without further information. But from various examples I see this can be all sorts of things, such as a list, key value pairs, python code returning a list or dict, or a mix thereof. From experimentation, it seems lists of functions that return lists are also ok --- so there's some flattening going on here too? It would be very nice to have a complete reference in mind.
* What's a `statement`? Below the grammar, it says "while all not defined non-terminals map to their Python equivalents". Does this mean that `statement` == "any Python statement"? Presumably this the Python in my current environment (`/usr/bin/env python`)? And what is the execution model here? Are different statements "independent"? How careful do I need to be about code with side effects? What about Python statements within parts of rules?
* What Python functions/variables are available to Snakefiles? I see things like `config`, `shell()`, `expand()` and `glob_wildcard()` in various examples. Of course I can `print(dir())` at the top of my Snakefile, but this is not a substitute for a comprehensive list of what functions are public and thus *supported*.

**Comment by epruesse** [comment]
Agreed. I ended up reading the source and putting `import pdb;pdb.set_trace()` into my snakefiles to figure out what I had and could do when I started with Snakemake.

What I would really like to see is a Snakemake 7.0 with some of the inconsistencies ironed out. The way the different parts of a rule are handled is quite different, and can be very confusing. E.g. `{something}` in `input:` would be a wildcard, while in `shell:` and `message` it would be replaced by the content of a local python variable. 

If you know Python, it will help to know that `input: something, else=bla` will be translated into `internalfunction(something, else=bla)` (it's a decorator, actually, for historic reasons).

The data passed as `input: something, else=bla` is stored as a `NamedList`, allowing you to access named parts with dot syntax (as attributes) as well as as array, etc. If you print it, print `something.__dict__`, so you can see the names and the content.

Functions are parsed with introspection. The first argument is always the wildcards, the others are parsed from the names, and which ones are available depends on where you use the function. Functions returning dicts need to be wrapped in `unpack()`.

**Comment by jlumpe** [comment]
Very much agreed, I recently opened [my own issue](#1168) on this but haven't gotten a response yet. I feel like this is a very essential part of the documentation that is omitted.
